### PR TITLE
(1/2) Implementing Compiler Pass for AutoTP

### PR DESCRIPTION
**Pull request by therealnaveenkamal**
Working on #8104 

- Added support for AutoTP. Added two primitives: copy_to_tp and reduce_from_tp - the f and g nodes.
-  Identifies column/row-parallel matmuls by the injected layer type in nn_module_stack, reading back.
- Wrote a test to verify correctness of module injection and compiler pass.

cc @tohtana

### CODE REVIEW

File: deepspeed/compile/custom_ops/tp_collectives.py [hunk 1]
```diff
@@ -0,0 +1,72 @@
+# Copyright (c) Microsoft Corporation.
```
Comment (tohtana) [inline deepspeed/compile/custom_ops/tp_collectives.py]: For new files, please show DeepSpeed's copyright. You can find an example in other files.

File: tests/unit/compile/test_tp_compile.py [hunk 2]
```diff
@@ -0,0 +1,183 @@
...
+
+import deepspeed
+import deepspeed.comm as dist
+from deepspeed.accelerator import get_accelerator
+from deepspeed.utils import groups
+from deepspeed.utils.torch import required_torch_version
+
+from unit.common import DistributedTest
+
+pytestmark = pytest.mark.skipif(not required_torch_version(min_version=2.9),
```
Comment (tohtana) [inline tests/unit/compile/test_tp_compile.py]: Can you clarify what limits this feature to v2.9+? Also, can we have the same check in the main code?

**Review by tohtana — CHANGES_REQUESTED** [review]
Thank you @therealnaveenkamal, this is amazing!
I don't see any critical issue in the code. The approach that reuses existing AutoTP's patterns is great. It gives consistent results with non-DeepCompile AutoTP.

One remaining work is validating the correctness in a more realistic setting. I think it would be good to compare loss values from  existing AutoTP and this one. I did similar work for [AutoEP](https://github.com/tohtana/DeepSpeedExamples/tree/tohtana/add_auto_ep/training/expert_parallel). The harness for the verification might be useful if you don't have such a script. We should check different configs like DP1/TP4 and DP2/TP2.

I left a few comments about details. Please consider addressing them. Also, please fix the commit to pass DCO check.

**Comment by tohtana** [comment]
Can you also share your plan for the the next step as this PR has 1/2 in the title.
I think it would be good to start with the combination with ZeRO. Probably we could do further optimizations based on profiling.

File: deepspeed/compile/passes/tp_compile.py [hunk 3]
```diff
@@ -0,0 +1,148 @@
...
+    return issubclass(layer_type, family)
+
+
+def defer_collectives_to_compiler(model) -> None:
+    """Suppress the module-level TP collectives on layers this pass will handle in the graph.
+
+    Any tensor-parallel layer the pass cannot rewrite is rejected rather than left on the
+    module-level path.
+    """
+    for name, module in model.named_modules():
```
Comment (tohtana) [inline deepspeed/compile/passes/tp_compile.py:54]: Can we set the flag `defer_collectives_to_compiler` only all modules passed the check?
If this raises an error in the loop, only some modules have `defer_collectives_to_compiler=True`. But the outer code might catch the error and fallback to eager. In that case, some communication collectives will be skipped.

File: tests/unit/module_inject/test_tp_plan_converter.py [hunk 4]
```diff
@@ -109,11 +108,25 @@ def test_pattern_matches_param_name(self):
...
+        wraps the MoE router, whose forward returns a tuple, and breaks the model outright.
+        """
         hf_plan = {"layers.*.q_proj": "local_colwise", "layers.*.o_proj": "rowwise"}
-        result = TPPlanConverter.convert(hf_plan)
-        assert result is None
+        specs = TPPlanConverter.convert(hf_plan)
+
+        assert specs is not None
+        by_type = {spec.partition_type for spec in specs}
+        assert PartitionType.ROW in by_type, "the supported entry should still be applied"
```
Comment (tohtana) [inline tests/unit/module_inject/test_tp_plan_converter.py]: I think this expectation (and the code) is wrong. To make TP work, COLUMN and ROW should be paired. Partially skipping the conversion breaks it.

Comment (tohtana) [reply tests/unit/module_inject/test_tp_plan_converter.py]: The issue with Llama4 MoE router should be addressed in another PR. How about making it fail when any unsupported style is found.

File: deepspeed/compile/passes/tp_compile.py [hunk 5]
```diff
@@ -0,0 +1,148 @@
...
+
+COLUMN_PARALLEL_OP = torch.ops.autotp.copy_to_tp_region.default
+ROW_PARALLEL_OP = torch.ops.autotp.reduce_from_tp_region.default
+GATHER_OUTPUT_OP = torch.ops.autotp.gather_from_tp_region.default
+
+# AutoTP replaces nn.Linear with these layers and shards their weights, so the injected layer type
+# already records the partitioning decision the pass needs. 
+
+COLUMN_PARALLEL_LAYERS = (LinearLayer, SubParamLinearLayer)
+ROW_PARALLEL_LAYERS = (LinearAllreduce, SubParamLinearAllreduce)
```
Comment (tohtana) [inline deepspeed/compile/passes/tp_compile.py:24]: SubParamLinearAllreduce.forward() always executes its module-level row all-reduce, while the compiler pass classifies that layer as row parallel and inserts another graph all-reduce.

**Review by tohtana — CHANGES_REQUESTED** [review]
Hi @therealnaveenkamal,
Thank you for the update! I think the earlier issues have been addressed.
I also found some correctness issues in the current code. Can you check them?

File: deepspeed/compile/passes/tp_compile.py (hunk 5 above)
Comment (therealnaveenkamal) [reply deepspeed/compile/passes/tp_compile.py:24]: thanks for this, @tohtana. I've made a check for `defer_collectives_to_compiler` in `layers.py`

File: deepspeed/compile/passes/tp_compile.py (hunk 3 above)
Comment (therealnaveenkamal) [reply deepspeed/compile/passes/tp_compile.py:54]: got it. now we go through all the modules and if any module is incompatible, we raise an error

File: tests/unit/module_inject/test_tp_plan_converter.py (hunk 4 above)
Comment (therealnaveenkamal) [reply tests/unit/module_inject/test_tp_plan_converter.py]: fixed it now.

**Comment by therealnaveenkamal** [comment]
Hi @tohtana - thanks for pointing out the bugs. I've made the changes. Let me know if this is okay. Thanks.

**Comment by tohtana** [comment]
Hi @therealnaveenkamal, thank you for the update!
Overall the change looks okay to me, but this test failed in my env. Can you check it?

```
tests/unit/compile/test_tp_compile.py::TestAutoTPCompileMoE::test_mixtral_matches_module_injection
```

**Comment by tohtana** [comment]
Hi @therealnaveenkamal,
I investigated the test failure. I think it was a bug on HF side that existed from v5.8.0 to v5.10.0.

It started with https://github.com/huggingface/transformers/pull/45621 and fixed by https://github.com/huggingface/transformers/pull/45634. I found you already pushed the version guard, but can we be more specific about versions?
I don't see any other issues. After we confirm the new version guard works, let's merge this PR.

**Comment by therealnaveenkamal** [comment]
Thanks @tohtana,  Guard is now exact: the range [5.8.0, 5.10.1) lives as a constant in init_tp.py, and the test skips it.

**Comment by tohtana** [comment]
Hi @therealnaveenkamal,
Great, thank you for the update!

I just found the code should work even with the affected versions as long as we choose other than `batched_mm`. I opened a small PR to your branch: https://github.com/therealnaveenkamal/DeepSpeed/pull/5/changes
Can you review it? Feel free to modifty it if you want.

**Comment by therealnaveenkamal** [comment]
@tohtana sorry about that. I've merged the PR.

**Review by tohtana — APPROVED** [review]
@therealnaveenkamal Thank you for your great work! This is definitely a significant step for DeepCompile.
I really appreciate your contribution to DeepSpeed.
